### ISSUE TITLE
Game of Life example stucks on slow machines and high speed settings

### ISSUE THREAD

**Issue by 0x7CFE** [iced-rs/iced#512]
I still do not quite understand the logic behind message generation, life `version`ing and `amount` to be `ticke`d. Looks like when large speed setting is set it causes subscription to generate too many `Tick` messages  which, in turn, causes them to be processed in random order causing out of sync issues.

That causes everything to just stop and no events are firing. I still don't quite get how that can happen but after that the subscription method is not called anymore and no messages are received.

It just sits still like this forever (but if I click Pause and then Play all starts running again, untill the next stall):
<img src="https://user-images.githubusercontent.com/3784017/92301876-8510f980-ef91-11ea-846f-5fe02e9284bc.png" width="50%" height="50%">

I noticed that when that happens, the last arm in the `update`'s  `match` statement starts firing:
```
                Message::Ticked {
                    result: Ok(life),
                    version,
                    tick_duration,
                } if version == self.version => {
                    self.state.update(life);
                    self.life_cache.clear();

                    self.version += 1;
                    self.last_tick_duration = tick_duration;
                }
                Message::Ticked {
                    result: Err(error), ..
                } => {
                    dbg!(error);
                }
                Message::Ticked { version, .. } => {
                    println!("version mismatch {}, {}", self.version, version);
                }
```

IMO, the whole thing should not rely on subscription evens. Rather, speed setting should directly map to an amount of game steps performed per tick, whereas subscription should just be treated as an UI refresh interval and be in a reasonable range, say, 10 per second.

P.S.: I'm posting this issue because I think,  this may actually be a more subtle bug that affects UI engine itself and not just a game logic.

**Comment by hecrj** [comment]
> Looks like when large speed setting is set it causes subscription to generate too many Tick messages which, in turn, causes them to be processed in random order causing out of sync issues.

The order of the `Tick` messages is irrelevant. Ticks are queued and computed in batches asynchronously. Only one batch is computed at once.

> That causes everything to just stop and no events are firing. I still don't quite get how that can happen but after that the subscription method is not called anymore and no messages are received.

The `subscription` method does not need to be called for the messages to happen. It has no side-effects.

> It just sits still like this forever 

I can reproduce the issue if I increase the `speed` limit to `1_000_000` and I run it in `debug` mode. I seem to have traced it all the way to `tokio`. The `Stream` produced by `futures::time::Every` seems to stop producing messages when using a very small interval in conjunction with `spawn_blocking`. The future does not panic, but simply seems to be delayed forever.

I cannot seem to reproduce the issue if I replace `tokio` with `async-std`. This seems to further confirm it is indeed a `tokio` issue (maybe related to https://github.com/tokio-rs/tokio/issues/2301).

> IMO, the whole thing should not rely on subscription evens. Rather, speed setting should directly map to an amount of game steps performed per tick, whereas subscription should just be treated as an UI refresh interval and be in a reasonable range, say, 10 per second.

Limiting the example to 10 frames per second, when most monitors can handle 60 or more is not desirable.

The speed is the amount of ticks per second. We want to display as many of these ticks in different frames, depending on the monitor refresh rate. The current setup adapts to target speed, framerate, and tick load.

The ideal solution here would be to use a subscription similar to [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) on the Web. We will eventually transition to it once such a subscription is available.

Additionally, the example is meant to be a stress test of sorts. The maximum speed is `1000`, which means that 1 message will be produced per millisecond. The runtime should be able to handle it without issues, especially when we consider `Application::update` is only taking a bunch of nanoseconds (as we can see in your screenshot).

In conclusion, we should investigate further what is going on with `tokio`.

**Comment by 0x7CFE** [comment]
@hecrj, thank you for the detailed reply!

Could you please tell me, what does it mean when, under normal conditions, we receive a `Message::Ticked` with incorrect version? I understand that this can happen if I reset the simulation while a future is running. But why this happens when high speed is selected?

**Comment by hecrj** [comment]
> But why this happens when high speed is selected?

I don't see it happening at high speed unless, as you said, I clear the grid while the previous batch is being computed. The `version` is only updated after a successful `Ticked` message, a `clear`, or a preset switch.

**Comment by 0x7CFE** [comment]
> I don't see it happening at high speed unless, as you said, I clear or switch the grid while a batch is being computed.

Well, my experience is quite different. In my case I receive dozens of mismatched versions without any user interaction. I think you can try adding some extra delay to the inner tick function to see if it will trigger weird behavior.

**Comment by hecrj** [comment]
> Well, my experience is quite different. In my case I receive dozens of mismatched versions without any user interaction. I think you can try adding some extra delay to the inner tick function to see if it will trigger weird behavior.

As I have already explained, the `tick` function taking more time should not cause any issues. For a specific board, only one tick batch is processed in the background at a given time.

It doesn't matter how much of a delay I add, processing a `Tick` message sets the `State::is_ticking` flag to `true` and this is not reset until a `Ticked` message happens (the batch finishes). Any further `Tick` messages while this flag is `true` will not spawn a new batch and, therefore, will not produce a `Ticked` message. Thus, the `version` should always match unless you clear or change the preset while the tick batch is being processed. There is no way for them to be out of order because there shouldn't be many of them to begin with!

Are you sure you are running the example without any modifications? What is the message log when you are getting dozens of mismatches? What do the versions look like?

**Comment by hecrj** [comment]
I have moved the `version` field from the `Grid` to to the `GameOfLife` struct in https://github.com/hecrj/iced/commit/4f2962d73f3bdeeca8a11817e404c45e91e2c2cc. This should express my previous points more explicitly.
